pubsub/query/query.go

```diff
@@ -0,0 +1,212 @@
+func New(s string) (*Query, error) {
+	p := &QueryParser{Buffer: "\"" + s + "\""}
+	p.Init()
+	if err := p.Parse(); err != nil {
+		return nil, err
+	}
+	return &Query{str: s, parser: p}, nil
+}
+
+// String returns an original string.
+func (q *Query) String() string {
+	return q.str
+}
+
+type operator uint8
+
+const (
+	opLessEqual operator = iota
+	opGreaterEqual
+	opLess
+	opGreater
+	opEqual
+	opContains
+)
+
+// Matches returns true if the query matches a given set of tags, false otherwise.
+//
+// For example, query "name=John" matches tags = {"name": "John"}. More
+// examples could be found in parser_test.go and query_test.go.
+func (q *Query) Matches(tags map[string]interface{}) bool {
+	if len(tags) == 0 {
+		return false
+	}
+
+	buffer, begin, end := q.parser.Buffer, 0, 0
+
+	var tag string
+	var op operator
+
+	for _, token := range q.parser.Tokens() {
+		switch token.pegRule {
+
+		case rulePegText:
+			begin, end = int(token.begin), int(token.end)
+		case ruletag:
+			tag = buffer[begin:end]
+		case rulele:
+			op = opLessEqual
+		case rulege:
+			op = opGreaterEqual
+		case rulel:
+			op = opLess
+		case ruleg:
+			op = opGreater
+		case ruleequal:
+			op = opEqual
+		case rulecontains:
+			op = opContains
+		case rulevalue:
+			if !match(tag, op, reflect.ValueOf(buffer[begin:end]), tags) {
+				return false
+			}
+		case rulenumber:
+			number := buffer[begin:end]
+			if strings.Contains(number, ".") {
+				value, err := strconv.ParseFloat(number, 32)
+				if err != nil {
+					panic(fmt.Sprintf("got %v. number format errors should be handled by the parser", err))
+				}
+				if !match(tag, op, reflect.ValueOf(value), tags) {
+					return false
+				}
+			} else {
+				value, err := strconv.Atoi(number)
+				if err != nil {
+					panic(fmt.Sprintf("got %v. number format errors should be handled by the parser", err))
+				}
+				if !match(tag, op, reflect.ValueOf(value), tags) {
+					return false
+				}
+			}
+		case ruletime:
+			value, err := time.Parse(time.RFC3339, buffer[begin:end])
+			if err != nil {
+				panic(fmt.Sprintf("got %v. time format errors should be handled by the parser", err))
+			}
+			if !match(tag, op, reflect.ValueOf(value), tags) {
+				return false
+			}
+		case ruledate:
+			value, err := time.Parse("2006-01-02", buffer[begin:end])
+			if err != nil {
+				panic(fmt.Sprintf("got %v. time format errors should be handled by the parser", err))
+			}
+			if !match(tag, op, reflect.ValueOf(value), tags) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func match(tag string, op operator, operand reflect.Value, tags map[string]interface{}) bool {
+	value, ok := tags[tag]
+	if !ok { // no tag in tags matching a tag from query
+		return false
+	}
+	switch operand.Kind() {
+	case reflect.Struct: // time
+		v, ok := value.(time.Time)
+		if !ok {
+			return false
+		}
+		switch op {
+		case opLessEqual:
+			return v.Before(operand.Interface().(time.Time)) || v.Equal(operand.Interface().(time.Time))
```

Comment:
can we put  `operand.Interface().(time.Time)` in a variable so we only need to write it once ? same for those below